src/Details.js

```diff
@@ -12,55 +12,72 @@ if (typeof window !== 'undefined') {
   require('details-element-polyfill')
 }
 
-const DetailsReset = styled('details')`
+const StyledDetails = styled('details')`
   & > summary {
     list-style: none;
   }
   & > summary::-webkit-details-marker {
     display: none;
   }
+
+  ${COMMON}
 `
 function getRenderer(children) {
   return typeof children === 'function' ? children : () => children
 }
 
-function DetailsBase({children, overlay, render = getRenderer(children), defaultOpen = false, ...rest}) {
-  const [open, setOpen] = useState(defaultOpen)
+function Details({
+  children,
+  overlay,
+  render = getRenderer(children),
+  open: userOpen,
+  onClickOutside,
+  onToggle,
+  defaultOpen = false,
+  ...rest
+}) {
+  const [internalOpen, setInternalOpen] = useState(defaultOpen)
   const ref = useRef(null)
+  // only use internal open state if user doesn't provide a value for the open prop
+  const open = userOpen || internalOpen
 
-  const closeMenu = useCallback(
+  const onClickOutsideInternal = useCallback(
     event => {
-      // only close the menu if we're clicking outside
-      if (event && event.target.closest('details') !== ref.current) {
-        setOpen(false)
-        document.removeEventListener('click', closeMenu)
+      if (event.target.closest('details') !== ref.current) {
+        onClickOutside && onClickOutside(event)
+        if (!event.defaultPrevented) {
+          setInternalOpen(false)
+        }
       }
     },
-    [ref]
+    [ref, onClickOutside, setInternalOpen]
   )
 
+  // handles the overlay behavior - closing the menu when clicking outside of it
   useEffect(() => {
-    if (overlay && open) {
-      document.addEventListener('click', closeMenu)
+    if (open && overlay) {
+      document.addEventListener('click', onClickOutsideInternal)
       return () => {
-        document.removeEventListener('click', closeMenu)
+        document.removeEventListener('click', onClickOutsideInternal)
       }
     }
-  }, [open, overlay, closeMenu])
+  }, [open, overlay, onClickOutsideInternal])
+
+  function handleToggle(e) {
+    onToggle && onToggle(e)
 
-  function toggle(event) {
-    setOpen(event.target.open)
+    if (!e.defaultPrevented) {
+      setInternalOpen(e.target.open)
```

Comment:
Right now we continue to manage the state whether or not the user is managing it as well, and we just don't pass the open prop down if the user provides it - does that feel ok @BinaryMuse? Or should we add to this if statement `&& !userManagingState`

Reply:
Without the error correcting effect, it probably makes sense not to manage our internal state because there's a change it'll get out of sync anyway, so I'd be fine with adding the condition.